compiler/noirc_frontend/src/parser/parser/expression.rs

```diff
@@ -0,0 +1,1708 @@
+        self.parse_expression_or_error_impl(false) // allow constructors
+    }
+
+    pub(crate) fn parse_expression_or_error_impl(
+        &mut self,
+        allow_constructors: bool,
+    ) -> Expression {
+        if let Some(expr) = self.parse_expression_impl(allow_constructors) {
+            expr
+        } else {
+            self.push_expected_expression();
+            Expression { kind: ExpressionKind::Error, span: self.span_at_previous_token_end() }
+        }
+    }
+
+    fn parse_expression_impl(&mut self, allow_constructors: bool) -> Option<Expression> {
+        self.parse_equal_or_not_equal(allow_constructors)
+    }
+
+    /// Term
+    ///    = UnaryOp Term
+    ///    | AtomOrUnaryRightExpression
+    pub(super) fn parse_term(&mut self, allow_constructors: bool) -> Option<Expression> {
+        let start_span = self.current_token_span;
+
+        if let Some(operator) = self.parse_unary_op() {
+            let Some(rhs) = self.parse_term(allow_constructors) else {
+                self.expected_label(ParsingRuleLabel::Expression);
+                return None;
+            };
+            let kind = ExpressionKind::prefix(operator, rhs);
+            let span = self.span_since(start_span);
+            return Some(Expression { kind, span });
+        }
+
+        self.parse_atom_or_unary_right(allow_constructors)
+    }
+
+    /// UnaryOp = '&' 'mut' | '-' | '!' | '*'
+    fn parse_unary_op(&mut self) -> Option<UnaryOp> {
+        if self.tokens_follow(Token::Ampersand, Token::Keyword(Keyword::Mut)) {
+            self.next_token();
+            self.next_token();
+            Some(UnaryOp::MutableReference)
+        } else if self.eat(Token::Minus) {
+            Some(UnaryOp::Minus)
+        } else if self.eat(Token::Bang) {
+            Some(UnaryOp::Not)
+        } else if self.eat(Token::Star) {
+            Some(UnaryOp::Dereference { implicitly_added: false })
+        } else {
+            None
+        }
+    }
+
+    /// AtomOrUnaryRightExpression
+    ///     = Atom
+    ///     | UnaryRightExpression
+    fn parse_atom_or_unary_right(&mut self, allow_constructors: bool) -> Option<Expression> {
+        let start_span = self.current_token_span;
+        let mut atom = self.parse_atom(allow_constructors)?;
+        let mut parsed;
+
+        loop {
+            (atom, parsed) = self.parse_unary_right(atom, start_span);
+            if parsed {
+                continue;
+            } else {
+                break;
+            }
+        }
+
+        Some(atom)
+    }
+
+    /// UnaryRightExpression
+    ///     = CallExpression
+    ///     | MemberAccessOrMethodCallExpression
+    ///     | CastExpression
+    ///     | IndexExpression
+    fn parse_unary_right(&mut self, mut atom: Expression, start_span: Span) -> (Expression, bool) {
+        let mut parsed;
+
+        (atom, parsed) = self.parse_call(atom, start_span);
+        if parsed {
+            return (atom, parsed);
+        }
```

Comment:
Why don't we return a `Option<Expression>` here?

Reply:
Because if we couldn't parse a unary we have to keep the old expression. I thought of some other ways to do this, like returning a struct... I don't like the end result but I couldn't find a better way to do it.